src/storage/src/storage/bidi/connector.rs

```diff
@@ -0,0 +1,842 @@
+
+    pub async fn reconnect(
+        &mut self,
+        status: tonic::Status,
+        ranges: Vec<ProtoRange>,
+    ) -> Result<(BidiReadObjectResponse, Connection<T::Stream>)> {
+        use crate::read_resume_policy::ReadResumePolicy;
+
+        let error = handle_redirect(self.spec.clone(), status);
+        self.reconnect_attempts += 1;
+        let policy = ResumeRedirect::new(self.options.read_resume_policy());
+        match policy.on_error(&ResumeQuery::new(self.reconnect_attempts), error) {
+            ResumeResult::Continue(_) => self.connect(ranges).await,
+            ResumeResult::Exhausted(e) => Err(e),
+            ResumeResult::Permanent(e) => Err(e),
+        }
+    }
+
+    async fn connect_attempt(
+        client: T,
+        spec: Arc<Mutex<BidiReadObjectSpec>>,
+        request: &BidiReadObjectRequest,
+        options: &RequestOptions,
+    ) -> Result<(BidiReadObjectResponse, Connection<T::Stream>)> {
+        let bucket_name = request
+            .read_object_spec
+            .as_ref()
+            .map(|s| s.bucket.as_str())
+            .unwrap_or_default();
+        if bucket_name
+            .strip_prefix("projects/_/buckets/")
+            .is_none_or(|x| x.is_empty())
+        {
+            use gax::error::binding::*;
+            let problem = SubstitutionFail::MismatchExpecting(
+                bucket_name.to_string(),
+                "projects/_/buckets/*",
+            );
+            let mismatch = SubstitutionMismatch {
+                field_name: "bucket",
+                problem,
+            };
+            let mismatch = PathMismatch {
+                subs: vec![mismatch],
+            };
+            let mismatch = BindingError {
+                paths: vec![mismatch],
+            };
+
+            return Err(crate::Error::binding(mismatch));
+        }
+        let x_goog_request_params = request
+            .read_object_spec
+            .iter()
+            .flat_map(|s| s.routing_token.iter())
+            .fold(format!("bucket={bucket_name}"), |s, token| {
+                s + &format!(",routing_token={token}")
+            });
+
+        let (tx, rx) = tokio::sync::mpsc::channel::<BidiReadObjectRequest>(100);
+        tx.send(request.clone()).await.map_err(Error::io)?;
+
+        let extensions = {
+            let mut e = tonic::Extensions::new();
+            e.insert(tonic::GrpcMethod::new(
+                "google.storage.v2.Storage",
+                "BidiReadObject",
+            ));
+            e
+        };
+        let path =
+            http::uri::PathAndQuery::from_static("/google.storage.v2.Storage/BidiReadObject");
+
+        let response = client
+            .start(
+                extensions,
+                path,
+                rx,
+                options,
+                &X_GOOG_API_CLIENT_HEADER,
+                &x_goog_request_params,
+            )
+            .await?;
+        Self::started(spec, tx, response).await
+    }
+
+    async fn started(
```

Comment:
nit: also could use a comment here, that this reads the first message from the stream.